pkg/cri/server/sandbox_stats_windows.go

```
@@ -19,17 +19,397 @@ package server
+	podSandboxStats.Attributes = &runtime.PodSandboxAttributes{
+		Id:          meta.ID,
+		Metadata:    meta.Config.GetMetadata(),
+		Labels:      meta.Config.GetLabels(),
+		Annotations: meta.Config.GetAnnotations(),
+	}
+
+	metrics, containers, err := c.listWindowsMetricsForSandbox(ctx, sandbox)
+	if err != nil {
+		return nil, fmt.Errorf("failed to obtain container stats during podSandboxStats call: %w", err)
+	}
+
+	statsMap, err := convertMetricsToWindowsStats(metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	podCPU, containerStats, err := c.toPodSandboxStats(sandbox, statsMap, containers, timestamp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert container stats during podSandboxStats call: %w", err)
+	}
+	podSandboxStats.Windows.Cpu = podCPU.Cpu
+	podSandboxStats.Windows.Memory = podCPU.Memory
+	podSandboxStats.Windows.Containers = containerStats
+
+	podSandboxStats.Windows.Network = windowsNetworkUsage(ctx, sandbox, timestamp)
+
+	pidCount, err := c.getSandboxPidCount(ctx, sandbox)
+	if err != nil {
+		return nil, err
+	}
+
+	podSandboxStats.Windows.Process = &runtime.WindowsProcessUsage{
+		Timestamp:    timestamp.UnixNano(),
+		ProcessCount: &runtime.UInt64Value{Value: pidCount},
+	}
+
+	c.saveSandBoxMetrics(podSandboxStats.Attributes.Id, podSandboxStats)
+
+	return podSandboxStats, nil
+}
+
+func convertMetricsToWindowsStats(metrics []*types.Metric) (map[string]*wstats.Statistics, error) {
+	statsMap := make(map[string]*wstats.Statistics)
+	for _, stat := range metrics {
+		var podRuntimeStats *wstats.Statistics
+		podStatsData, err := typeurl.UnmarshalAny(stat.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract container metrics: %w", err)
+		}
+
+		// extract the metrics if available
+		// podStatsData can be nil for pods that don't have an actual podsandbox
+		// this is the case for hostprocess pods
+		if podStats, ok := podStatsData.(*wstats.Statistics); ok {
```

Suggest we log again if the type assertion fails. e.g. if the stats type changed, would be good to see that in logs as an error instead of silently failing.

---

I guess we should either throw errors in both failure cases in this function, or log in both, depending on if the overall pod stats operation should continue.

---

logging this makes sense to catch the off chance that it changed, but since it is an expected in the case of HPC this isn't a hard error. The previous failure is a hard failure so we should end early.

---

Based on @dcantah feedback, I've wired through the config to tell if this is a HPC, so I can make this a hard error.